internal/engine/upper.go

```diff
@@ -41,35 +43,38 @@ const maxChangedFilesToPrint = 5
 // TODO(nick): maybe this should be called 'BuildEngine' or something?
 // Upper seems like a poor and undescriptive name.
 type Upper struct {
-	b            BuildAndDeployer
-	watcherMaker watcherMaker
-	timerMaker   timerMaker
-	k8s          k8s.Client
-	browserMode  BrowserMode
-	reaper       build.ImageReaper
-	hud          hud.HeadsUpDisplay
+	b               BuildAndDeployer
+	fsWatcherMaker  fsWatcherMaker
+	timerMaker      timerMaker
+	podWatcherMaker podWatcherMaker
+	k8s             k8s.Client
+	browserMode     BrowserMode
+	reaper          build.ImageReaper
+	hud             hud.HeadsUpDisplay
 }
 
-type watcherMaker func() (watch.Notify, error)
+type fsWatcherMaker func() (watch.Notify, error)
+type podWatcherMaker func(context.Context, k8s.Client) (*podWatcher, error)
 type timerMaker func(d time.Duration) <-chan time.Time
 
 func NewUpper(ctx context.Context, b BuildAndDeployer, k8s k8s.Client, browserMode BrowserMode,
 	reaper build.ImageReaper, hud hud.HeadsUpDisplay) Upper {
-	watcherMaker := func() (watch.Notify, error) {
+	fsWatcherMaker := func() (watch.Notify, error) {
 		return watch.NewWatcher()
 	}
 
 	// Run the HUD in the background
 	go hud.Run(ctx)
 
 	return Upper{
-		b:            b,
-		watcherMaker: watcherMaker,
-		timerMaker:   time.After,
-		k8s:          k8s,
-		browserMode:  browserMode,
-		reaper:       reaper,
-		hud:          hud,
+		b:               b,
+		fsWatcherMaker:  fsWatcherMaker,
+		podWatcherMaker: makePodWatcher,
```

Comment:
can we use wire.go to inject this?

in theory, this shouldn't need to take a k8s.Client, because wire.go is supposed to take care of that.

Reply:
yep, thanks